source/common/network/connection_impl.cc

```diff
@@ -921,6 +921,18 @@ bool ConnectionImpl::bothSidesHalfClosed() {
   return read_end_stream_ && write_end_stream_ && write_buffer_->length() == 0;
 }
 
+bool ConnectionImpl::setSocketOption(Network::Socket::OptionConstSharedPtr option) {
+  if (socket_ == nullptr) {
+    return false;
+  }
+  if (option == nullptr) {
+    return false;
+  }
+  return option->setOption(
+      *socket_,
+      envoy::config::core::v3::SocketOption::SocketState::SocketOption_SocketState_STATE_LISTENING);
```

Comment:
I don't think it makes sense to use STATE_LISTENING here. This is always an established connection.

Also, whatever value you use here, you need to document that a passed in option with a different `in_state` won't be applied, but will return `true`.

It would be a nicer interface if you made a socket option that doesn't require a specific in_state.

Reply:
I added a STATE_ANY option and updated the socket implementation to always attempt to set the option regardless of the state of the socket. This should allow for the desired behavior without a new socket implementation or function.

WDYT?

Reply:
I updated the approach to allow for the socket option to be created without a specific `in_state`.